lib/request_started_on_middleware.rb

```diff
@@ -0,0 +1,57 @@
+# Add to config/application.rb:
+#
+#     config.middleware.use 'RequestStartedOnMiddleware'
+#
+class RequestStartedOnMiddleware
+  def initialize(app)
+    @app = app
+  end
+
+  def call(env)
+    start_request(env['PATH_INFO'], Time.now.utc)
+    @app.call(env)
+  ensure
+    complete_request
+  end
+
+  def start_request(path, started_on)
+    Thread.current[:current_request] = path
+    Thread.current[:current_request_started_on] = started_on
+  end
+
+  def complete_request
+    Thread.current[:current_request] = nil
+    Thread.current[:current_request_started_on] = nil
+  end
+
+  def self.long_running_requests
+    requests = []
+    timed_out_request_started_on = request_timeout.ago.utc
+
+    relevant_thread_list.each do |thread|
+      request    = thread[:current_request]
+      started_on = thread[:current_request_started_on]
+
+      # There's a race condition where the complete_request method runs in another
+      # thread after we set one or more of the above local variables. The fallout
+      # of this is we return a false positive for a request that finished very close
+      # to the 2 minute timeout.
+      if request.present? && started_on.kind_of?(Time) && timed_out_request_started_on > started_on
+        duration = (Time.now.utc - started_on).to_f
+        requests << [request, duration, thread]
+      end
+    end
+
+    requests
+  end
+
+  REQUEST_TIMEOUT = 2.minutes
```

Comment:
Two quick suggestions about this (which I didn't think to add into my last review):

* Maybe rename this to `LONG_REQUEST`, or something, since we really aren't "timing out" anything here.  It is based on the apache timeout, but this simply logs a long request, and this is the value for what constitutes that.
* Possibly make this only `1.minute`?  On the performance team, we were told to try and make anything over 10 seconds in the UI faster, so waiting until 2.minutes to say something is "long" seems like it is a bit extreme.  When paired with my suggestion to possibly include the `.backtrace` with it, it seems like this would allow us to more quickly see when something is long and causing issues, and if the issue is moving from one code path to the next.  And most requests (ideally) shouldn't get to this point where it is a problem.

Reply:
I agree on both.  I like the `LONG_REQUEST` name... it makes more sense.  I was thinking we could expose the `2.minutes` value because I also see value in changing it.  Maybe `1.minute` across the board will be better as it will pick up long requests that don't currently timeout apache but also requests that do.  I'll change it to `1.minute` and we'll go from there.